### ISSUE TITLE
Request Headers support is needed in DownloadRequest

### ISSUE THREAD

**Issue by nikky12** [androidx/media#359]
Before filing a feature request:
-----------------------
- Search existing open issues, specifically with the label ‘enhancement’:
  https://github.com/androidx/media/labels/enhancement
- For ExoPlayer-related feature requests, please also check for existing feature
  requests on the ExoPlayer tracker:
  https://github.com/google/ExoPlayer/labels/enhancement
- Search existing pull requests:
    - On this tracker: https://github.com/androidx/media/pulls,
    - On the ExoPlayer tracker: https://github.com/google/ExoPlayer/pulls

When filing a feature request:
-----------------------
Replace the content in the sections below.

### [REQUIRED] Use case description
We are using authenticated playback url. For fetching the segments we need to send some headers in requests. Same goes for content download and offline playback.
For downloading the content we are making use of DownloadRequest class. 

- This class does not have any field for holding request headers.
- DownloadRequest is a final class so can not be extended.

### Proposed solution
Can DownloadRequest class be made open for extension. So that we can make our custom requests like adding headers or request parameters map etc.

### Alternatives considered
A clear and concise description of any alternative solutions you considered,
if applicable.

**Comment by marcbaechinger** [comment]
The `DownloadService` needs to setup a `DownloadManager`. Does it help if you set this up with a `ResoilvingDataSource` somehow? You can then set the headers when the download starts and change the data spec accordingly.

```
DefaultHttpDataSource.Factory upstreamFactory = new DefaultHttpDataSource.Factory();
new DataSource.Factory() {
  @Override
  public DataSource createDataSource() {
    return new ResolvingDataSource(
        upstreamFactory.createDataSource(),
        dataSpec ->
            dataSpec.buildUpon().setHttpRequestHeaders(ImmutableMap.of("foo", "bar")).build());
  }
};
DownloadManager downloadManager =
    new DownloadManager(
        context,
        getDatabaseProvider(context),
        getDownloadCache(context),
        getHttpDataSourceFactory(context),
        Executors.newFixedThreadPool(/* nThreads= */ 6));
```

https://developer.android.com/guide/topics/media/exoplayer/customization#customizing_server_interactions
https://developer.android.com/reference/androidx/media3/datasource/ResolvingDataSource

**Comment by nikky12** [comment]
@marcbaechinger We are creating DataSource.Factory and downloadManager only once for all downloads. May you plls elaborate how can we set the headers when the download starts and change the data spec accordingly.

Here is our code snippet
downloadManager = DownloadManager(
            context,
            databaseProvider,
            downloadCache,
            dataSourceFactory,
            downloadExecutor
        )
After setting downloadManager for DownloadService we are calling below api
DownloadService.sendAddDownload(
                context,
                service,
                downloadRequest,
                /* foreground= */ true
            )

**Comment by marcbaechinger** [comment]
Please read the documentation links I sent above. I'm not going to repeat what is written there I'm afraid.

The `Resolver` that you pass to the constructor of `ResolvingDataSource` is called each time when a request is done. You can change the `DataSpec` for each request.

**Comment by marcbaechinger** [comment]
I'm closing this issue due to inactivity and because I think that using a `ResolvingDataSource` can be used to set headers. Please shout or reopen a new issue if this isn't the case.
